interp/interp_test.go

```diff
@@ -191,3 +194,228 @@ func TestPiecewiseConstantPredict(t *testing.T) {
 	leftYs := []float64{-0.5, 1.5, 1.5, 1.5, 1, 1}
 	testInterpolatorPredict(t, pc, testXs, leftYs, 0)
 }
+
+func TestPiecewiseCubicPredict(t *testing.T) {
+	t.Parallel()
+	xs := []float64{-1, 0, 1}
+	lastY := rightPoly(xs[2])
+	coeffs := mat.NewDense(2, 4, []float64{3, -3, 1, 0, 1, -1, 0, 1})
+	pc := PiecewiseCubic{xs, coeffs, lastY}
+	testFittedPolys(t, &pc)
+}
+
+func TestPiecewiseCubicFitWithDerivatives(t *testing.T) {
+	t.Parallel()
+	xs := []float64{-1, 0, 1}
+	ys := make([]float64, 3)
+	dydxs := make([]float64, 3)
+	ys[0] = leftPoly(xs[0])
+	ys[1] = leftPoly(xs[1])
+	ys[2] = rightPoly(xs[2])
+	dydxs[0] = leftPolyDerivative(xs[0])
+	dydxs[1] = leftPolyDerivative(xs[1])
+	dydxs[2] = rightPolyDerivative(xs[2])
+	var pc PiecewiseCubic
+	err := pc.FitWithDerivatives(xs, ys, dydxs)
+	if err != nil {
+		t.Errorf("Error when fitting piecewise cubic interpolator: %v", err)
+	}
+	testFittedPolys(t, &pc)
+	lastY := rightPoly(xs[2])
+	if pc.lastY != lastY {
+		t.Errorf("Mismatch in lastY: got %v, want %g", pc.lastY, lastY)
+	}
+	if !floats.Equal(pc.xs, xs) {
+		t.Errorf("Mismatch in xs: got %v, want %v", pc.xs, xs)
+	}
+	coeffs := mat.NewDense(2, 4, []float64{3, -3, 1, 0, 1, -1, 0, 1})
+	if !mat.EqualApprox(pc.coeffs, coeffs, 1e-14) {
+		t.Errorf("Mismatch in coeffs: got %v, want %v", pc.coeffs, coeffs)
+	}
+}
+
+func TestPiecewiseCubicFitWithDerivativesErrors(t *testing.T) {
+	t.Parallel()
+	for _, test := range []struct {
+		xs, ys, dydxs   []float64
+		expectedMessage string
+	}{
+		{
+			xs:              []float64{0, 1, 2},
+			ys:              []float64{10, 20},
+			dydxs:           []float64{0, 0, 0},
+			expectedMessage: differentLengths,
+		},
+		{
+			xs:              []float64{0, 1, 1},
+			ys:              []float64{10, 20, 30},
+			dydxs:           []float64{0, 0, 0, 0},
+			expectedMessage: differentLengths,
+		},
+		{
+			xs:              []float64{0},
+			ys:              []float64{0},
+			dydxs:           []float64{0},
+			expectedMessage: tooFewPoints,
+		},
+		{
+			xs:              []float64{0, 1, 1},
+			ys:              []float64{10, 20, 10},
+			dydxs:           []float64{0, 0, 0},
+			expectedMessage: xsNotStrictlyIncreasing,
+		},
+	} {
+		var pc PiecewiseCubic
+		err := pc.FitWithDerivatives(test.xs, test.ys, test.dydxs)
+		if err == nil || err.Error() != test.expectedMessage {
+			t.Errorf("expected error for xs: %v, ys: %v and dydxs: %v with message: %s", test.xs, test.ys, test.dydxs, test.expectedMessage)
+		}
+	}
+}
+
+func leftPoly(x float64) float64 {
+	return x*x - x + 1
+}
+
+func leftPolyDerivative(x float64) float64 {
+	return 2*x - 1
+}
+
+func rightPoly(x float64) float64 {
+	return x*x*x - x + 1
+}
+
+func rightPolyDerivative(x float64) float64 {
+	return 3*x*x - 1
+}
+
+func testFittedPolys(t *testing.T, pc *PiecewiseCubic) {
```

Comment:
The testing for Cubic should be structured differently. For example, this function works only if `pc` was fitted to particular points using particular functions (`rightPoly` and `leftPoly`). This hard-wired connection is surprising, also because the name `testFittedPolys` looks quite general. Also, the package-level functions `rightPoly`, `leftPoly` are not very informative.

I would have a table of test cases, something like you have for Akima:
```
type testCaseCubic struct {
    x []float64
    y []float64
    dy []float64

    want func(float64) float64
}

for _, test := range []testCaseCubic {
    {
        x: []float64{0, 0.9},
        y: []float64{0, 0},
        dy: []float64{0, 0},
        want: func(float64) float64 {return 0},
    },
    {
        x: []float64{0, 0.9, 2.3},
        y: []float64{1, 1, 1},
        dy: []float64{0, 0, 0},
        want: func(float64) float64 {return 1},
    },
    ...
}
```
You can have helper functions to create test cases:
```
func newTestCaseCubic(n int, f, df func(float64) float64, rnd *rand.Rand) testCaseCubic {
    x := make([]float64, n)
    // fill x, for example by jittering uniformly spaced points
    y := make([]float64, n)
    dy := make([]float64, n)
    for i, xi := range x {
        y[i] = f(xi)
        dy[i] = df(xi)
    }
    return testCaseCubic {
        x: x,
        y: y,
        dy: dy,
        want: f,
    }
}
```
In this way you can have a variety of interesting test cases, from some special "corner cases" like when the cubic is actually a constant or linear function to a general piecewise cubic. The input data and the expected output are grouped together, so the test code is easier to document and follow.

Reply:
I cleaned it up.